### PR TITLE
[AIR] <Part 3> Add LightningPredictor to support batch prediction

### PR DESCRIPTION
**Pull request by woshiyyya**
## Why are these changes needed?

There will be a list of PRs for PyTorch Lightning Integration. This is the second one.

- LightningTrainer + Test
- Logging and Checkpointing + Test
- **Predictor + Batch Prediction + Test**
- Release Tests + Docstring tests

The `LightningPredictor` is a subclass of `TorchPredictor`. They shares the same interfaces for predictor and batchPrediction.

Content for this PR:
- [x] Implemented LightningPredictor, loading from LightningCheckpoint
- [x] Tested simple prediction with different fs and different input format
- [x] Tested BatchPrediction compatibility

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: python/ray/train/tests/test_lightning_predictor.py [hunk 1]
```diff
@@ -0,0 +1,139 @@
...
+def save_checkpoint(model: pl.LightningModule, ckpt_path: str):
+    trainer = pl.Trainer(max_epochs=0)
+    trainer.fit(model, train_dataloaders=DataLoader(torch.randn(1)))
+    trainer.save_checkpoint(ckpt_path)
+
+
+@pytest.mark.parametrize("fs", ["s3", "local"])
+@pytest.mark.parametrize("use_gpu", [True, False])
+@pytest.mark.parametrize("use_preprocessor", [True, False])
+@pytest.mark.parametrize("batch_format", ["numpy", "pandas"])
```
Comment (justinvyu) [inline python/ray/train/tests/test_lightning_predictor.py:39]: Maybe we don't need to test `batch_format` and leave it to Data unit tests. Just choose numpy

File: python/ray/train/tests/test_lightning_predictor.py [hunk 2]
```diff
@@ -0,0 +1,139 @@
...
+    batch_format: str,
+):
+    model_config = {
+        "layer_1": 32,
+        "layer_2": 64,
+        "lr": 1e-4,
+    }
+    model = LightningMNISTClassifier(**model_config)
+
+    with tempfile.TemporaryDirectory() as tmpdir:
```
Comment (justinvyu) [inline python/ray/train/tests/test_lightning_predictor.py:54]: Tip: You can use the `tmp_path` pytest fixture.

```python
def test_predictor(tmp_path, ...):
    ckpt_path = str(tmp_path / MODEL_KEY(
```

It returns a pathlib `Path`.

File: python/ray/train/tests/test_lightning_predictor.py [hunk 3]
```diff
@@ -0,0 +1,139 @@
...
+        ckpt_path = f"{tmpdir}/{MODEL_KEY}"
+        save_checkpoint(model, ckpt_path)
+
+        # Test load checkpoint from local dir or remote path
+        checkpoint = LightningCheckpoint.from_path(ckpt_path)
+        if fs == "s3":
+            checkpoint.to_uri(mock_s3_bucket_uri)
+            checkpoint = LightningCheckpoint.from_uri(mock_s3_bucket_uri)
```
Comment (justinvyu) [inline python/ray/train/tests/test_lightning_predictor.py:62]: Should we test `from_directory` as well? Aka if they try loading from an AIR checkpoint directory.

Then, we can pytest search over:

```python
"checkpoint_source": ["from_path", "from_uri", "from_directory"]
```

File: python/ray/train/tests/test_lightning_predictor.py [hunk 4]
```diff
@@ -0,0 +1,139 @@
...
+
+        preprocessor = DummyPreprocessor() if use_preprocessor else None
+
+        # Instantiate a predictor from checkpoint
+        predictor = LightningPredictor.from_checkpoint(
+            checkpoint=checkpoint,
+            model=LightningMNISTClassifier,
+            use_gpu=use_gpu,
```
Comment (justinvyu) [inline python/ray/train/tests/test_lightning_predictor.py:70]: Any way to check GPU is actually used?

File: python/ray/train/tests/test_lightning_predictor.py [hunk 5]
```diff
@@ -0,0 +1,139 @@
...
+@pytest.mark.parametrize("use_gpu", [True, False])
+@pytest.mark.parametrize("use_preprocessor", [True, False])
+@pytest.mark.parametrize("batch_format", ["numpy", "pandas"])
+def test_predictor(
+    mock_s3_bucket_uri,
+    fs: str,
+    use_preprocessor: bool,
+    use_gpu: bool,
+    batch_format: str,
+):
```
Comment (justinvyu) [inline python/ray/train/tests/test_lightning_predictor.py:46]: 1 line docstring explaining high level what is done in this test and what the assertions/requirements are at the end?

File: python/ray/train/lightning/lightning_predictor.py [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+@PublicAPI(stability="alpha")
+class LightningPredictor(TorchPredictor):
+    """A predictor for PyTorch Lightning modules.
+
+    Args:
+        model: The PyTorch Lightning module to use for predictions.
+        preprocessor: A preprocessor used to transform data batches prior
+            to prediction.
+        use_gpu: If set, the model will be moved to GPU on instantiation and
+            prediction happens on GPU.
```
Comment (justinvyu) [inline python/ray/train/lightning/lightning_predictor.py:23]: Can we add a tested code example in this docstring?

File: python/ray/train/lightning/lightning_predictor.py [hunk 7]
```diff
@@ -0,0 +1,64 @@
...
+        use_gpu: bool = False,
+    ):
+        super(LightningPredictor, self).__init__(
+            model=model, preprocessor=preprocessor, use_gpu=use_gpu
+        )
+
+    @classmethod
+    def from_checkpoint(
+        cls,
+        checkpoint: Checkpoint,
```
Comment (justinvyu) [inline python/ray/train/lightning/lightning_predictor.py:39]: Should the type be a `LightningCheckpoint`?

File: python/ray/train/lightning/lightning_predictor.py [hunk 8]
```diff
@@ -0,0 +1,64 @@
...
+            model_class: A subclass of ``pytorch_lightning.LightningModule`` that
+                defines your model and training logic.
+            preprocessor: A preprocessor used to transform data batches prior
+                to prediction.
+            use_gpu: If set, the model will be moved to GPU on instantiation and
+                prediction happens on GPU.
+            **load_from_checkpoint_kwargs: Arguments to pass into
+                ``pl.LightningModule.load_from_checkpoint``
+        """
+        checkpoint = LightningCheckpoint.from_checkpoint(checkpoint)
```
Comment (justinvyu) [inline python/ray/train/lightning/lightning_predictor.py:62]: Should we be using this? When will a user pass in a generic `Checkpoint` into this? I think we can require `LightningCheckpoint`.

**Review by justinvyu — CHANGES_REQUESTED** [review]
Looks good, some suggestions:

File: python/ray/train/tests/test_lightning_predictor.py (hunk 1 above)
Comment (woshiyyya) [reply python/ray/train/tests/test_lightning_predictor.py:39]: ok

File: python/ray/train/lightning/lightning_predictor.py (hunk 6 above)
Comment (woshiyyya) [reply python/ray/train/lightning/lightning_predictor.py:68]: Added!

File: python/ray/train/tests/test_lightning_predictor.py (hunk 5 above)
Comment (woshiyyya) [reply python/ray/train/tests/test_lightning_predictor.py:47]: Done!

File: python/ray/train/lightning/lightning_predictor.py (hunk 7 above)
Comment (woshiyyya) [reply python/ray/train/lightning/lightning_predictor.py]: Same below

File: python/ray/train/lightning/lightning_predictor.py [hunk 9]
```diff
@@ -0,0 +1,111 @@
...
+                defines your model and training logic. Note that this is a class type
+                instead of a model instance.
+            preprocessor: A preprocessor used to transform data batches prior
+                to prediction.
+            use_gpu: If set, the model will be moved to GPU on instantiation and
+                prediction happens on GPU.
+            **load_from_checkpoint_kwargs: Arguments to pass into
+                ``pl.LightningModule.load_from_checkpoint``
+        """
+        model = checkpoint.get_model(model, **load_from_checkpoint_kwargs)
```
Comment (gjoliver) [inline python/ray/train/lightning/lightning_predictor.py:110]: maybe you don't need to unpack then pack. just pass it like:
```
model = checkpoint.get_model(
    model, load_from_checkpoint_kwargs=load_from_checkpoint_kwargs
)
```
will do?

**Review by gjoliver — APPROVED** [review]
looks cool. just 1 minor nit.

File: python/ray/train/lightning/lightning_predictor.py (hunk 9 above)
Comment (woshiyyya) [reply python/ray/train/lightning/lightning_predictor.py]: Good idea. Updated!

**Review by justinvyu — APPROVED** [review]
✅

**Pull request merged**